### PR TITLE
MTD geometry: replace string with string_view in MTDBaseNumber 

### PR DESCRIPTION
**Pull request by fabiocos**
#### PR description:

Following recent discussions triggered by simulation performance reports, this PR update the ```MTDBaseNumber``` class, moving it from the use of ```string``` to the use of ```string_view```. This is possible since the object is just used locally to build a ```DetId``` and the underlying source of strings (geometry volume names) is always available to the code while being invoked. 

All the classes using this tool are updated accordingly. 

#### PR validation:

Unit tests pass, and the workflow 34634.0 runs successfully without any change in DQM output.
An igprof test on the GEN-SIM part of the workflow (10 events) show a reduction of the rank of the ```getBaseNumber``` method used inside ```MtdSD``` (already a minor contribution anyway):

 Original
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
            0.0  .........       0.10 / 0.10         MtdSD::setDetUnitId(G4Step const*) [1743]
[1748]      0.0       0.10       0.00 / 0.10       MtdSD::getBaseNumber(G4Step const*)
            0.0  .........       0.08 / 0.08         MTDBaseNumber::addLevel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int const&) [1882]
            0.0  .........       0.02 / 0.02         MTDBaseNumber::reset() [2969]
```

Updated
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
            0.0  .........       0.04 / 0.06         MtdSD::setDetUnitId(G4Step const*) [2020]
[2294]      0.0       0.04       0.00 / 0.04       MtdSD::getBaseNumber(G4Step const*)
            0.0  .........       0.02 / 0.02         MTDBaseNumber::reset() [2988]
            0.0  .........       0.01 / 12.08        G4TouchableHistory::GetVolume(int) const [112]
            0.0  .........       0.01 / 9.97         _init [122]
```

A similar approach could be used for ```EcalBaseNumber```, which has a much higher impact on the GEN-SIM step performances.

### CODE REVIEW

**Comment by fabiocos** [comment]
please test

File: Geometry/MTDCommonData/interface/MTDBaseNumber.h [hunk 1]
```diff
@@ -19,20 +19,20 @@ class MTDBaseNumber {
   ~MTDBaseNumber() {}
 
   void setSize(const int& size);
-  void addLevel(const std::string& name, const int& copyNumber);
+  void addLevel(const std::string_view& name, const int& copyNumber);
```
Comment (makortel) [inline Geometry/MTDCommonData/interface/MTDBaseNumber.h:22]: The `std::string_view`s are generally recommended to be passed around by value.

File: Geometry/MTDCommonData/src/BTLNumberingScheme.cc [hunk 2]
```diff
@@ -44,18 +44,19 @@ uint32_t BTLNumberingScheme::getUnitID(const MTDBaseNumber& baseNumber) const {
...
-      ::atoi((baseName.substr(8, 1)).c_str()) == 9 || ::atoi((baseName.substr(8, 1)).c_str()) == 5 ||
-              (::atoi((baseName.substr(8, 1)).c_str()) == 7 && ::atoi((baseName.substr(7, 1)).c_str()) == 1)
-          ? ::atoi((baseName.substr(7, 1)).c_str()) + 1
-          : ::atoi((baseName.substr(7, 1)).c_str()));
+  int tmptyp = ::atoi(&baseName.at(7));
+  if (tmptyp == 17) {
+    tmptyp = 2;
+  } else if (tmptyp == 33) {
+    tmptyp = 3;
+  }
```
Comment (cvuosalo) [inline Geometry/MTDCommonData/src/BTLNumberingScheme.cc:58]: What about the cases where the second digit is 5 or 9? It seems the value should be incremented by 1.

File: Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc [hunk 3]
```diff
@@ -310,12 +310,12 @@ void DD4hep_TestMTDIdealGeometry::theBaseNumber(cms::DDFilteredView& fv) {
   thisN_.setSize(fv.navPos().size());
 
   for (uint ii = 0; ii < fv.navPos().size(); ii++) {
-    std::string name((fv.geoHistory()[ii])->GetName());
-    name.assign(name.erase(name.rfind('_')));
-    int copyN(fv.copyNos()[ii]);
-    thisN_.addLevel(name, copyN);
+    std::string_view name((fv.geoHistory()[ii])->GetName());
+    size_t ipos = name.rfind('_');
+    thisN_.addLevel((static_cast<std::string_view>((fv.geoHistory()[ii])->GetName())).substr(0, ipos),
```
Comment (cvuosalo) [inline Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc:315]:
```suggestion
    thisN_.addLevel((name.substr(0, ipos),
```

Comment (cvuosalo) [reply Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc:315]: Two more cases of this line below should also be changed.

File: Geometry/MTDCommonData/src/BTLNumberingScheme.cc (hunk 2 above)
Comment (fabiocos) [reply Geometry/MTDCommonData/src/BTLNumberingScheme.cc:58]: @cvuosalo the other cases were referred to old BTL geometry versions that are no more available in recent releases.
I do not see the point to maintain that with a more complex logic.

File: Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc (hunk 3 above)
Comment (fabiocos) [reply Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc]: given the nature of ```string_view``` it works even if ```name``` is local, I agree

File: Geometry/MTDCommonData/interface/MTDBaseNumber.h (hunk 1 above)
Comment (fabiocos) [reply Geometry/MTDCommonData/interface/MTDBaseNumber.h]: thanks for the comment, updated

**Comment by fabiocos** [comment]
please test

**Comment by cvuosalo** [comment]
+1

**Comment by fabiocos** [comment]
@AdrianoDee @srimanob any comment?

**Comment by srimanob** [comment]
+Upgrade

Technical PR to improve performance. PR test runs fine with no change in DQM.

**Comment by perrotta** [comment]
+1

**Pull request merged**